### PR TITLE
Overhaul quicklisp-to-nix

### PR DESCRIPTION
**Pull request by SquircleSpace**
###### Motivation for this change

The quicklisp-to-nix system misses many dependencies and is awkward to run.  Some lisp systems cannot be loaded because they weren't compiled before the nix package directory was made readonly.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by bradleyjensen** [comment]
Note: I took the (potentially?) contentious stance that lispPackages should be about lisp code and we shouldn't be building the executables provided by lisp systems.  I believe that executables provided by lispPackages (such as stumpwm) should be in their own nix package.

**Comment by Mic92** [comment]
Makes sense. We also move out applications out of `pythonPackages`.

**Comment by 7c6f434c** [comment]
We don't have any Lisp binaries that you are expected to run without writing additional Lisp code, as far as I know…

**Comment by bradleyjensen** [comment]
I can't reproduce the build failure for 7a9dxzcii40616l2dlv1vkjnkf21n501-lisp-alexandria-20170630-git on my NixOS box.  I did the following.

1. Added "nix.useSandbox = true;" to my system's configuration.nix
2. Ran "nixos-rebuild test"
3. Removed all gc roots related to my local clone of nixpkgs
4. Merged my branch with top-of-tree master
5. Built alexandria.  It built cleanly with the same hash.

Am I missing something?

File: pkgs/development/lisp-modules/define-package.nix [hunk 1]
```diff
@@ -43,11 +46,34 @@ let
     echo "export LD_LIBRARY_PATH=\"\$NIX_LISP_LD_LIBRARY_PATH\''${NIX_LISP_LD_LIBRARY_PATH:+:}\$LD_LIBRARY_PATH\"" >> "$launch_script"
     echo '"${clwrapper}/bin/common-lisp.sh" "$@"' >> "$launch_script"
   '';
+  moveAsdFiles = ''
+    find $out/lib/common-lisp/ -name '*.asd' | while read ASD_FILE; do
+      KEEP_THIS_ASD=0
+      for ALLOWED_ASD in $asdFilesToKeep; do
+        ALLOWED_ASD="/$ALLOWED_ASD"
+        ALLOWED_ASD_LENGTH=${"$"}{#ALLOWED_ASD}
```
Comment (7c6f434c) [inline pkgs/development/lisp-modules/define-package.nix]: By the way, do you know about escaping by `''`, i.e. `'' x=''${y}''`?

File: pkgs/development/lisp-modules/define-package.nix [hunk 2]
```diff
@@ -43,11 +46,34 @@ let
...
+        ASD_FILE_LENGTH=${"$"}{#ASD_FILE}
+        ASD_FILE_SUFFIX_INDEX=$(expr "$ASD_FILE_LENGTH" - "$ALLOWED_ASD_LENGTH")
+        ASD_FILE_SUFFIX_INDEX=$(expr "$ASD_FILE_SUFFIX_INDEX" + 1)
+        echo $ALLOWED_ASD $ASD_FILE $ASD_FILE_SUFFIX_INDEX $(expr substr "$ASD_FILE" "$ASD_FILE_SUFFIX_INDEX" "$ASD_FILE_LENGTH")
+        if [ "$(expr substr "$ASD_FILE" "$ASD_FILE_SUFFIX_INDEX" "$ASD_FILE_LENGTH")" == "$ALLOWED_ASD" ]; then
+          KEEP_THIS_ASD=1
+          break
+        fi
```
Comment (7c6f434c) [inline pkgs/development/lisp-modules/define-package.nix]: Is all the computation to check if `test "$ALLOWED_ASD" = "$(basename "ASD_FILE")"`?

File: pkgs/development/lisp-modules/quicklisp-to-nix-overrides.nix [hunk 3]
```diff
@@ -23,9 +22,6 @@ in
         cp "$out/lib/common-lisp/stumpwm/stumpwm" "$out/bin"
       '';
     };
-    propagatedBuildInputs = (x.propagatedBuildInputs or []) ++ (with qlnp; [
-      alexandria cl-ppcre clx
-    ]);
   };
   iterate = skipBuildPhase;
```
Comment (7c6f434c) [inline pkgs/development/lisp-modules/quicklisp-to-nix-overrides.nix]: If I understand your changes correctly, `skipBuildPhase` is now the default

**Comment by 7c6f434c** [comment]
On the whole great, thanks.

**Comment by 7c6f434c** [comment]
Did you run the package update without using sandboxing?

**Comment by bradleyjensen** [comment]
I did not know about that quoting behavior. Thanks!

I don’t know why I didn’t think to use basename. Brain fart. Sorry about that.

Correct. Skipping Rhee build is now the default. I suppose I should have removed the unnecessary overrides.

**Comment by 7c6f434c** [comment]
I dunno if it would make sense to keep going to list _all_ the systems that couldn't be processed. Unfortunately, even among Quicklisp-provided systems, that is not always the empty set.

**Comment by bradleyjensen** [comment]
I’m not sure if I follow. Are you saying we should try to get every buildable quicklisp system into lispPackages? Last time I checked, there were about a thousand releases and about 3 thousand quicklispable systems provided by those releases.

**Comment by 7c6f434c** [comment]
No, I mean that when I add multiple systems (or even try to switch to the next upstream version) before regenerating, handling some systems fails, and not only because of the missing dependencies.

**Pull request merged**